ott/geometry/costs.py

```diff
@@ -122,27 +124,111 @@ def __init__(self, dimension: int, **kwargs: Any):
     self._dimension = dimension
     self._sqrtm_kw = kwargs
 
-  def norm(self, x: jnp.ndarray) -> Union[jnp.ndarray, float]:
-    norm = jnp.sum(x[..., 0:self._dimension] ** 2, axis=-1)
-    x_mat = jnp.reshape(
-        x[..., self._dimension:], (-1, self._dimension, self._dimension)
-    )
-
-    norm += jnp.trace(x_mat, axis1=-2, axis2=-1)
+  def norm(self, x):
+    mean, cov = self.x_to_mean_and_cov(x)
+    norm = jnp.sum(mean ** 2, axis=-1)
+    norm += jnp.trace(cov, axis1=-2, axis2=-1)
     return norm
 
-  def pairwise(self, x: jnp.ndarray, y: jnp.ndarray) -> float:
-    mean_dot_prod = jnp.vdot(x[0:self._dimension], y[0:self._dimension])
-    x_mat = jnp.reshape(x[self._dimension:], (self._dimension, self._dimension))
-    y_mat = jnp.reshape(y[self._dimension:], (self._dimension, self._dimension))
-
-    sq_x = matrix_square_root.sqrtm(x_mat, self._dimension, **self._sqrtm_kw)[0]
-    sq_x_y_sq_x = jnp.matmul(sq_x, jnp.matmul(y_mat, sq_x))
+  def pairwise(self, x, y):
+    mean_x, cov_x = self.x_to_mean_and_cov(x)
+    mean_y, cov_y = self.x_to_mean_and_cov(y)
+    mean_dot_prod = jnp.vdot(mean_x, mean_y)
+    sq_x = matrix_square_root.sqrtm(cov_x, self._dimension, **self._sqrtm_kw)[0]
+    sq_x_y_sq_x = jnp.matmul(sq_x, jnp.matmul(cov_y, sq_x))
     sq__sq_x_y_sq_x = matrix_square_root.sqrtm(
         sq_x_y_sq_x, self._dimension, **self._sqrtm_kw
     )[0]
     return -2 * (mean_dot_prod + jnp.trace(sq__sq_x_y_sq_x, axis1=-2, axis2=-1))
 
+  def x_to_mean_and_cov(self, x):
+    x = jnp.atleast_2d(x)
+    means = x[:, 0:self._dimension]
+    covariances = jnp.reshape(
+        x[:, self._dimension:self._dimension + self._dimension ** 2],
+        (-1, self._dimension, self._dimension)
+    )
+    return jnp.squeeze(means), jnp.squeeze(covariances)
+
+  @functools.partial(jax.vmap, in_axes=[None, None, 0, 0])
+  def scale_covariances(self, cov_sqrt, cov_i, lambda_i):
+    return lambda_i * matrix_square_root.sqrtm_only(
+        jnp.matmul(jnp.matmul(cov_sqrt, cov_i), cov_sqrt)
+    )
+
+  def relative_diff(self, x: jnp.ndarray, y: jnp.ndarray) -> jnp.ndarray:
+    return jnp.sum(jnp.square(x - y)) / jnp.prod(jnp.array(x.shape))
+
+  def covariance_fixpoint_iter(
+      self,
+      covs: jnp.ndarray,
+      lambdas: jnp.ndarray,
+      rtol: float = 1e-2
+  ) -> jnp.ndarray:
+
+    def cond_fn(iteration, constants, state):
+      _, diff = state
+      return diff > jnp.array(rtol)
+
+    def body_fn(iteration, constants, state, compute_error):
+      del compute_error
+      cov, _ = state
+      cov_sqrt = matrix_square_root.sqrtm_only(cov)
+      scaled_cov = jnp.linalg.matrix_power(
+          jnp.sum(self.scale_covariances(cov_sqrt, covs, lambdas), axis=0), 2
+      )
+      cov_sqrt_minus = jnp.linalg.matrix_power(cov_sqrt, -1)
```

Comment:
if you also need the inverse of cov_sqrt, why not use directly [`sqrtm`](https://github.com/ott-jax/ott/blob/4de67b8aaea241742f3be3385b02ecbad1350911/ott/geometry/matrix_square_root.py#L29) directly? this returns the inverse as well for free. Minor detail: maybe better call `cov_sqrt_minus` the more accurate `cov_inv_sqrt`